crates/red_knot_test/src/db.rs

```diff
@@ -1,69 +1,60 @@
-use std::sync::Arc;
-
+use camino::{Utf8Component, Utf8PathBuf};
 use red_knot_python_semantic::lint::{LintRegistry, RuleSelection};
-use red_knot_python_semantic::{
-    default_lint_registry, Db as SemanticDb, Program, ProgramSettings, PythonPlatform,
-    SearchPathSettings,
-};
+use red_knot_python_semantic::{default_lint_registry, Db as SemanticDb};
 use ruff_db::files::{File, Files};
-use ruff_db::system::{DbWithTestSystem, System, SystemPath, SystemPathBuf, TestSystem};
+use ruff_db::system::{
+    DbWithWritableSystem, InMemorySystem, OsSystem, System, SystemPath, SystemPathBuf,
+    WritableSystem,
+};
 use ruff_db::vendored::VendoredFileSystem;
 use ruff_db::{Db as SourceDb, Upcast};
-use ruff_python_ast::PythonVersion;
+use ruff_notebook::{Notebook, NotebookError};
+use std::borrow::Cow;
+use std::sync::Arc;
+use tempfile::TempDir;
 
 #[salsa::db]
 #[derive(Clone)]
 pub(crate) struct Db {
-    project_root: SystemPathBuf,
     storage: salsa::Storage<Self>,
     files: Files,
-    system: TestSystem,
+    system: MdtestSystem,
     vendored: VendoredFileSystem,
     rule_selection: Arc<RuleSelection>,
 }
 
 impl Db {
-    pub(crate) fn setup(project_root: SystemPathBuf) -> Self {
+    pub(crate) fn setup() -> Self {
         let rule_selection = RuleSelection::from_registry(default_lint_registry());
 
-        let db = Self {
-            project_root,
+        Self {
+            system: MdtestSystem::in_memory(SystemPathBuf::from("/")),
             storage: salsa::Storage::default(),
-            system: TestSystem::default(),
             vendored: red_knot_vendored::file_system().clone(),
             files: Files::default(),
             rule_selection: Arc::new(rule_selection),
-        };
-
-        db.memory_file_system()
-            .create_directory_all(&db.project_root)
-            .unwrap();
-
-        Program::from_settings(
-            &db,
-            ProgramSettings {
-                python_version: PythonVersion::default(),
-                python_platform: PythonPlatform::default(),
-                search_paths: SearchPathSettings::new(vec![db.project_root.clone()]),
-            },
-        )
-        .expect("Invalid search path settings");
-
-        db
+        }
     }
 
-    pub(crate) fn project_root(&self) -> &SystemPath {
-        &self.project_root
+    pub(crate) fn reset_fs(&mut self) {
+        match &*self.system.0 {
+            MdtestSystemInner::InMemory(in_memory) => {
+                in_memory.fs().remove_all();
+            }
+            MdtestSystemInner::Os { .. } => {
+                self.system = MdtestSystem::in_memory(SystemPathBuf::from("/"));
+            }
+        }
+
+        Files::sync_all(self);
     }
-}
 
-impl DbWithTestSystem for Db {
-    fn test_system(&self) -> &TestSystem {
-        &self.system
+    pub(crate) fn with_tempdir_fs(&mut self, cwd: SystemPathBuf, temp_dir: TempDir) {
+        self.system.with_os(cwd, temp_dir);
```

Comment:
I think the `with_*` naming is surprising for methods that mutate in-place; that name strongly suggests an immutable builder pattern where `self` is cloned and the updated one is returned. (Because in that case the method name describes the returned object.)

For in-place mutation I would expect e.g. `set_tempdir_fs` and `set_os` instead.

Reply:
>  that name strongly suggests an immutable builder pattern where self is cloned and the updated one is returned

I've seen both usages of `with_` where they take `self` or `&mut self`. I'd say it's even more common for builders taking `self` to omit the `with_` prefix altogether (because they don't have getters).

Reply:
I think the more relevant distinction maybe is that I would expect builder methods named `with_*` to return `self` (whether they clone or mutate in place) so they can be called chained; again because the `with_*` naming describes what is returned from the method.